optaplanner-core/src/main/java/org/optaplanner/core/impl/util/Sequence.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.util;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class Sequence<T> {
```

Comment:
This is public API, as it is exported by a public method. I suggest we treat these public types in the following way:

- Interface with the user-facing stuff is in `api`. (`Sequence`)
- Implementation with constructors and all the protected stuff is in `impl`. (`SequenceImpl`)
- Implementations have every method public, but the classes themselves are not.

This way, we make it very clear what is or is not accessible by users, but we get to hide the implementation.
Please apply the same to `Break`, `ConsecutiveData` and any other type that actually reaches the user.